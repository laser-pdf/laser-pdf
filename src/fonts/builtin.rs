use std::{collections::HashMap, str::CharIndices};

use afm::{CharMetric, FontMetrics, Ligature};
use pdf_core_14_font_afms::*;
use pdf_writer::Name;

use super::{EncodedGlyph, Font, ShapedGlyph};
use crate::Pdf;

pub struct BuiltinFont {
    resource_name: String,
    metrics: FontMetrics,
    char_metrics_by_codepoint: HashMap<u32, CharMetric>,
}

impl BuiltinFont {
    fn add(pdf: &mut Pdf, font_name: &str, afm: &str) -> Self {
        let parser = afm::afm();

        let mut input = pom::DataInput::new(afm.as_bytes());

        let metrics = parser.parse(&mut input).unwrap();

        let mut char_metrics_by_codepoint = HashMap::new();

        for char_metric in &metrics.char_metrics {
            char_metrics_by_codepoint.insert(
                char_metric.character_code as u32,
                // some manual cloning
                CharMetric {
                    name: char_metric.name.clone(),
                    ligatures: char_metric
                        .ligatures
                        .iter()
                        .map(|l| Ligature {
                            ligature: l.ligature.clone(),
                            successor: l.successor.clone(),
                            ..*l
                        })
                        .collect(),
                    ..*char_metric
                },
            );
        }

        let id = pdf.alloc();

        pdf.pdf
            .type1_font(id)
            .base_font(pdf_writer::Name(font_name.as_bytes()));

        let idx = pdf.fonts.len();
        pdf.fonts.push(id);

        let resource_name = format!("F{}", idx);

        BuiltinFont {
            resource_name,
            metrics,
            char_metrics_by_codepoint,
        }
    }

    pub fn courier(pdf: &mut Pdf) -> Self {
        Self::add(pdf, "Courier", COURIER)
    }

    pub fn courier_bold(pdf: &mut Pdf) -> Self {
        Self::add(pdf, "Courier-Bold", COURIER_BOLD)
    }

    pub fn courier_oblique(pdf: &mut Pdf) -> Self {
        Self::add(pdf, "Courier-Oblique", COURIER_OBLIQUE)
    }

    pub fn courier_bold_oblique(pdf: &mut Pdf) -> Self {
        Self::add(pdf, "Courier-BoldOblique", COURIER_BOLD_OBLIQUE)
    }

    pub fn helvetica(pdf: &mut Pdf) -> Self {
        Self::add(pdf, "Helvetica", HELVETICA)
    }

    pub fn helvetica_bold(pdf: &mut Pdf) -> Self {
        Self::add(pdf, "Helvetica-Bold", HELVETICA_BOLD)
    }

    pub fn helvetica_oblique(pdf: &mut Pdf) -> Self {
        Self::add(pdf, "Helvetica-Oblique", HELVETICA_OBLIQUE)
    }

    pub fn helvetica_bold_oblique(pdf: &mut Pdf) -> Self {
        Self::add(pdf, "Helvetica-BoldOblique", HELVETICA_BOLD_OBLIQUE)
    }

    pub fn times_roman(pdf: &mut Pdf) -> Self {
        Self::add(pdf, "Times-Roman", TIMES_ROMAN)
    }

    pub fn times_bold(pdf: &mut Pdf) -> Self {
        Self::add(pdf, "Times-Bold", HELVETICA_BOLD)
    }

    pub fn times_italic(pdf: &mut Pdf) -> Self {
        Self::add(pdf, "Times-Italic", TIMES_ITALIC)
    }

    pub fn times_bold_italic(pdf: &mut Pdf) -> Self {
        Self::add(pdf, "Times-BoldItalic", TIMES_BOLD_ITALIC)
    }

    pub fn symbol(pdf: &mut Pdf) -> Self {
        Self::add(pdf, "Symbol", SYMBOL)
    }

    pub fn zapf_dingbats(pdf: &mut Pdf) -> Self {
        Self::add(pdf, "ZapfDingbats", ZAPF_DINGBATS)
    }
}

#[derive(Clone)]
pub struct Shaped<'a> {
    font: &'a BuiltinFont,
    chars: CharIndices<'a>,
}

impl<'a> Iterator for Shaped<'a> {
    type Item = ShapedGlyph;

    fn next(&mut self) -> Option<Self::Item> {
        self.chars.next().map(|(i, c)| {
            let metrics = self.font.char_metrics_by_codepoint.get(&(c as u32));

            let advance = metrics.map_or(0, |m| m.wx as i32); // i hope those are ints in practice

            ShapedGlyph {
                unsafe_to_break: false,
                glyph_id: c as u32, // i guess
                text_range: i..i + c.len_utf8(),
                x_advance_font: advance,
                x_advance: advance,
                x_offset: 0,
                y_offset: 0,
                y_advance: 0,
            }
        })
    }
}

impl Font for BuiltinFont {
    type Shaped<'a>
        = Shaped<'a>
    where
        Self: 'a;

    fn shape<'a>(&'a self, text: &'a str, _: i32, _: i32) -> Self::Shaped<'a> {
        Shaped {
            font: self,
            chars: text.char_indices(),
        }
    }

    fn encode(&self, _: &mut crate::Pdf, glyph_id: u32, _: &str) -> EncodedGlyph {
        EncodedGlyph::OneByte(glyph_id as u8)
    }

    fn resource_name(&self) -> pdf_writer::Name {
        Name(self.resource_name.as_bytes())
    }

    fn units_per_em(&self) -> u16 {
        1000
    }

    fn general_metrics(&self) -> super::GeneralMetrics {
        let bbox = self.metrics.font_bbox;

        // This should be bbox.ymax - bbox.ymin, but it seems that the afm is parsed incorrectly.
        let line_height = bbox.ymax - bbox.xmax;

        let ascent = line_height + self.metrics.descender;

        super::GeneralMetrics {
            ascent: ascent as u32,
            line_height: line_height as u32,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_no_panic() {
        let mut pdf = Pdf::new();

        BuiltinFont::courier(&mut pdf);
        BuiltinFont::courier_bold(&mut pdf);
        BuiltinFont::courier_oblique(&mut pdf);
        BuiltinFont::courier_bold_oblique(&mut pdf);

        BuiltinFont::helvetica(&mut pdf);
        BuiltinFont::helvetica_bold(&mut pdf);
        BuiltinFont::helvetica_oblique(&mut pdf);
        BuiltinFont::helvetica_bold_oblique(&mut pdf);

        BuiltinFont::times_roman(&mut pdf);
        BuiltinFont::times_bold(&mut pdf);
        BuiltinFont::times_italic(&mut pdf);
        BuiltinFont::times_bold_italic(&mut pdf);

        BuiltinFont::symbol(&mut pdf);
        BuiltinFont::zapf_dingbats(&mut pdf);
    }
}
