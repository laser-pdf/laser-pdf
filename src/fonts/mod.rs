pub mod builtin;
pub mod truetype;

use std::ops::Range;

pub struct GeneralMetrics {
    pub ascent: u32,
    pub line_height: u32,
}

#[derive(Debug, Clone)]
pub struct ShapedGlyph {
    pub unsafe_to_break: bool,
    pub glyph_id: u32,
    pub text_range: Range<usize>,
    /// without kerning
    pub x_advance_font: i32,
    pub x_advance: i32,
    pub x_offset: i32,
    pub y_advance: i32,
    pub y_offset: i32,
}

// TODO: different representation?
pub enum EncodedGlyph {
    OneByte(u8),
    TwoBytes([u8; 2]),
}

pub trait Font {
    type Shaped<'a>: Iterator<Item = ShapedGlyph> + Clone + 'a
    where
        Self: 'a;

    fn shape<'a>(
        &'a self,
        text: &'a str,
        character_spacing: i32,
        word_spacing: i32,
    ) -> Self::Shaped<'a>;

    fn encode(&self, pdf: &mut crate::Pdf, glyph_id: u32, text: &str) -> EncodedGlyph;

    fn resource_name(&self) -> pdf_writer::Name<'_>;

    fn general_metrics(&self) -> GeneralMetrics;

    fn units_per_em(&self) -> u16;
}
