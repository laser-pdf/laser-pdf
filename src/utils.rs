use pdf_writer::Content;
// use printpdf::*;

// pub fn circle(layer: &PdfLayerReference, pos: [f32; 2], radius: f32) {
//     let circle = printpdf::utils::calculate_points_for_circle(Pt(radius), Pt(pos[0]), Pt(pos[1]));

//     layer.add_shape(Line {
//         points: circle,
//         is_closed: true,
//         has_fill: true,
//         has_stroke: false,
//         is_clipping_path: false,
//     });
// }

pub fn line(layer: &mut Content, pos: (f32, f32), width: f32, thickness: f32) {
    // TODO: mm_to_pt
    layer
        .set_line_width(thickness)
        .move_to(mm_to_pt(pos.0), mm_to_pt(pos.1))
        .line_to(mm_to_pt(pos.0 + width), mm_to_pt(pos.1))
        .stroke();
}

pub fn mm_to_pt(mm: f32) -> f32 {
    // (mm as f64 * 2.834_646) as f32
    // (mm as f64 * 72. / 25.4) as f32
    mm * 72. / 25.4
}

pub fn pt_to_mm(pt: f32) -> f32 {
    // (pt as f64 * 0.352_778) as f32
    pt * 25.4 / 72.
}

pub fn u32_to_color_and_alpha(color: u32) -> ([f32; 3], f32) {
    (
        [
            ((color & 0xff_00_00_00) >> 24) as f32 / 255.0,
            ((color & 0x00_ff_00_00) >> 16) as f32 / 255.0,
            ((color & 0x00_00_ff_00) >> 8) as f32 / 255.0,
        ],
        (color & 0x00_00_00_ff) as f32 / 255.0,
    )
}

pub fn u32_to_rgb_color_array(color: u32) -> [u8; 3] {
    [
        ((color & 0xff_00_00_00) >> 24) as u8,
        ((color & 0x00_ff_00_00) >> 16) as u8,
        ((color & 0x00_00_ff_00) >> 8) as u8,
    ]
}

pub fn rgb_color_array_to_u32(color: [u8; 3]) -> u32 {
    ((color[0] as u32) << 24) | ((color[1] as u32) << 16) | ((color[2] as u32) << 8) | 0xFF
}

pub fn max_optional_size(a: Option<f32>, b: Option<f32>) -> Option<f32> {
    match (a, b) {
        (None, None) => None,
        (None, Some(x)) | (Some(x), None) => Some(x),
        (Some(a), Some(b)) => Some(a.max(b)),
    }
}

pub fn add_optional_size(a: Option<f32>, b: Option<f32>) -> Option<f32> {
    match (a, b) {
        (None, None) => None,
        (None, Some(x)) | (Some(x), None) => Some(x),
        (Some(a), Some(b)) => Some(a + b),
    }
}

pub fn add_optional_size_with_gap(a: Option<f32>, b: Option<f32>, gap: f32) -> Option<f32> {
    match (a, b) {
        (None, None) => None,
        (None, Some(x)) | (Some(x), None) => Some(x),
        (Some(a), Some(b)) => Some(a + gap + b),
    }
}

pub fn set_fill_color(layer: &mut Content, color: u32) {
    let (color, _) = u32_to_color_and_alpha(color);
    layer.set_fill_rgb(color[0], color[1], color[2]);
}

pub fn set_stroke_color(layer: &mut Content, color: u32) {
    let (color, _) = u32_to_color_and_alpha(color);
    layer.set_stroke_rgb(color[0], color[1], color[2]);
}

pub fn scale(scale: f32) -> [f32; 6] {
    [scale, 0., 0., scale, 0., 0.]
}
