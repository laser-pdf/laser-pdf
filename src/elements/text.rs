use fonts::GeneralMetrics;
use text::{Line, Lines, draw_line, lines};
use utils::{mm_to_pt, pt_to_mm, set_fill_color};

use crate::{fonts::Font, *};

#[derive(Copy, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub enum TextAlign {
    Left,
    Center,
    Right,
}

/// A text element that renders text content with various styling options.
pub struct Text<'a, F: Font> {
    /// The text content to render
    pub text: &'a str,
    /// Font reference
    pub font: &'a F,
    /// Font size in points
    pub size: f32,
    /// Text color as RGBA (default: black 0x00_00_00_FF)
    pub color: u32,
    /// Whether to underline the text
    pub underline: bool,
    /// Additional spacing between characters
    pub extra_character_spacing: f32,
    /// Additional spacing between words
    pub extra_word_spacing: f32,
    /// Additional line height
    pub extra_line_height: f32,
    /// Text alignment
    pub align: TextAlign,
}

struct FontMetrics {
    ascent: f32,
    line_height: f32,
}

impl<'a, F: Font> Text<'a, F> {
    pub fn basic(text: &'a str, font: &'a F, size: f32) -> Self {
        Text {
            text,
            font,
            size,
            color: 0x00_00_00_FF,
            underline: false,
            extra_character_spacing: 0.,
            extra_word_spacing: 0.,
            extra_line_height: 0.,
            align: TextAlign::Left,
        }
    }

    fn compute_font_metrics(&self) -> FontMetrics {
        let GeneralMetrics {
            ascent,
            line_height,
        } = self.font.general_metrics();

        let units_per_em = self.font.units_per_em() as f32;

        FontMetrics {
            ascent: pt_to_mm(ascent as f32 * self.size / units_per_em),
            line_height: pt_to_mm(line_height as f32 * self.size / units_per_em)
                + self.extra_line_height,
        }
    }

    #[inline(always)]
    fn render_lines<'b, L: Iterator<Item = Line<F::Shaped<'a>>>>(
        &self,
        text: &str,
        lines: L,
        mut ctx: DrawCtx,
        ascent: f32,
        line_height: f32,
        width: f32,
    ) -> (f32, f32) {
        let mut max_width = width;
        let mut last_line_full_width = 0;

        let mut x = ctx.location.pos.0;
        let mut y = ctx.location.pos.1 - ascent;

        let mut height_available = ctx.first_height;

        let mut line_count = 0;
        let mut draw_rect = 0;

        let start = |pdf: &mut Pdf, location: &Location| {
            let layer = location.layer(pdf);
            layer
                .save_state()
                .set_font(self.font.resource_name(), self.size);
            set_fill_color(layer, self.color);
            layer.begin_text();
        };

        let end = |pdf: &mut Pdf, location: &Location| {
            location.layer(pdf).end_text().restore_state();
        };

        start(ctx.pdf, &ctx.location);

        for line in lines {
            let line_width =
                pt_to_mm(line.width as f32 / self.font.units_per_em() as f32 * self.size);
            max_width = max_width.max(line_width);

            last_line_full_width = line.width + line.trailing_whitespace_width;

            if height_available < line_height {
                if let Some(ref mut breakable) = ctx.breakable {
                    end(ctx.pdf, &ctx.location);

                    let new_location = (breakable.do_break)(
                        ctx.pdf,
                        draw_rect,
                        if line_count == 0 {
                            None
                        } else {
                            Some(line_count as f32 * line_height)
                        },
                    );
                    draw_rect += 1;
                    x = new_location.pos.0;
                    y = new_location.pos.1 - ascent;
                    height_available = breakable.full_height;
                    ctx.location.page_idx = new_location.page_idx;
                    ctx.location.layer_idx = new_location.layer_idx;
                    line_count = 0;

                    start(ctx.pdf, &ctx.location);
                }
            }

            let layer = ctx.location.layer(ctx.pdf);

            let x_offset = match self.align {
                TextAlign::Left => 0.,
                TextAlign::Center => (width - line_width) / 2.,
                TextAlign::Right => width - line_width,
            };

            let x = x + x_offset;

            layer.set_text_matrix([1.0, 0.0, 0.0, 1.0, mm_to_pt(x), mm_to_pt(y)]);

            draw_line(ctx.pdf, &ctx.location, self.font, text, line);

            y -= line_height;
            height_available -= line_height;
            line_count += 1;
        }

        end(ctx.pdf, &ctx.location);

        (
            max_width.max(pt_to_mm(
                last_line_full_width as f32 / self.font.units_per_em() as f32 * self.size,
            )),
            line_count as f32 * line_height,
        )
    }

    #[inline(always)]
    fn layout_lines<'b, L: Iterator<Item = Line<F::Shaped<'a>>>>(
        &self,
        lines: L,
        line_height: f32,
        measure_ctx: Option<&mut MeasureCtx>,
    ) -> (f32, f32) {
        let mut max_width = 0;
        let mut last_line_full_width = 0;
        let mut line_count = 0;

        // This function is a bit hacky because it's both used for measure and for determining the
        // max line width in unconstrained-width contexts.
        let mut height_available = if let Some(&mut MeasureCtx { first_height, .. }) = measure_ctx {
            first_height
        } else {
            f32::INFINITY
        };

        for line in lines {
            if let Some(&mut MeasureCtx {
                breakable: Some(ref mut breakable),
                ..
            }) = measure_ctx
            {
                if height_available < line_height {
                    *breakable.break_count += 1;
                    height_available = breakable.full_height;
                    line_count = 0;
                }
            }

            max_width = max_width.max(line.width);
            last_line_full_width = line.width + line.trailing_whitespace_width;

            height_available -= line_height;
            line_count += 1;
        }

        (
            pt_to_mm(
                max_width.max(last_line_full_width) as f32 / self.font.units_per_em() as f32
                    * self.size,
            ),
            line_count as f32 * line_height,
        )
    }

    fn break_into_lines<R>(
        &'a self,
        width: f32,
        f: impl for<'b> FnOnce(Lines<'a, 'b, F>) -> R,
    ) -> R {
        lines(
            self.font,
            (self.extra_character_spacing / self.size * self.font.units_per_em() as f32) as i32,
            (self.extra_word_spacing / self.size * self.font.units_per_em() as f32) as i32,
            // The ceil is to prevent rounding errors from causing problems in cases where the
            // element gets measured and then the measured width gets used for draw, such as in
            // HAlign.
            (mm_to_pt(width) / self.size * self.font.units_per_em() as f32).ceil() as u32,
            self.text,
            f,
        )
    }
}

impl<'a, F: Font> Element for Text<'a, F> {
    fn first_location_usage(&self, ctx: FirstLocationUsageCtx) -> FirstLocationUsage {
        let FontMetrics {
            ascent: _,
            line_height,
        } = self.compute_font_metrics();

        if line_height > ctx.first_height {
            FirstLocationUsage::WillSkip
        } else {
            FirstLocationUsage::WillUse
        }
    }

    fn measure(&self, mut ctx: MeasureCtx) -> ElementSize {
        let FontMetrics { line_height, .. } = self.compute_font_metrics();

        let size = self.break_into_lines(ctx.width.max, |lines| {
            self.layout_lines(lines, line_height, Some(&mut ctx))
        });

        ElementSize {
            width: Some(ctx.width.constrain(size.0)),
            height: Some(size.1),
        }
    }

    fn draw(&self, ctx: DrawCtx) -> ElementSize {
        let FontMetrics {
            ascent,
            line_height,
        } = self.compute_font_metrics();

        // For left alignment we don't need to pre-layout because the
        // x offset is always zero.
        let width = if ctx.width.expand {
            ctx.width.max
        } else if self.align == TextAlign::Left {
            0.
        } else {
            // TODO: Figure out a way to avoid shaping twice here.
            self.break_into_lines(ctx.width.max, |lines| {
                self.layout_lines(lines, line_height, None).0
            })
        };

        let width_constraint = ctx.width;
        let size = self.break_into_lines(ctx.width.max, |lines| {
            self.render_lines(self.text, lines, ctx, ascent, line_height, width)
        });

        ElementSize {
            width: Some(width_constraint.constrain(size.0)),
            height: Some(size.1),
        }
    }
}

#[cfg(test)]
mod tests {
    use elements::column::Column;
    use fonts::truetype::TruetypeFont;
    use insta::*;

    use crate::test_utils::binary_snapshots::*;
    use crate::{
        DrawCtx, ElementSize,
        fonts::builtin::BuiltinFont,
        test_utils::{ElementProxy, ElementTestParams},
    };

    use super::*;

    const FONT: &[u8] = include_bytes!("../fonts/Kenney Bold.ttf");

    #[test]
    fn test_multi_page() {
        let bytes = test_element_bytes(TestElementParams::breakable(), |mut callback| {
            let font = BuiltinFont::courier(callback.pdf());

            let content = Text::basic(LOREM_IPSUM, &font, 32.);
            let content = content.debug(0);

            callback.call(&content);
        });
        assert_binary_snapshot!(".pdf", bytes);
    }

    #[test]
    fn test_truetype() {
        let bytes = test_element_bytes(TestElementParams::breakable(), |mut callback| {
            let font = TruetypeFont::new(callback.pdf(), FONT);

            let content = Text::basic(LOREM_IPSUM, &font, 32.);
            let content = content.debug(0);

            callback.call(&content);
        });
        assert_binary_snapshot!(".pdf", bytes);
    }

    #[test]
    fn test_truetype_trailing_whitespace() {
        let mut params = TestElementParams::breakable();
        params.width.expand = false;

        let bytes = test_element_bytes(params, |mut callback| {
            let font = TruetypeFont::new(callback.pdf(), FONT);

            let content = Text::basic("Whitespace ", &font, 32.);
            let content = content.debug(0);

            callback.call(&content);
        });
        assert_binary_snapshot!(".pdf", bytes);
    }

    #[test]
    fn test_truetype_extra_spacing() {
        let mut params = TestElementParams::breakable();
        params.width.expand = false;

        let bytes = test_element_bytes(params, |mut callback| {
            let font = TruetypeFont::new(callback.pdf(), FONT);

            callback.call(&Column {
                gap: 12.,
                collapse: false,
                content: |content| {
                    let normal = Text::basic("Hello, World", &font, 32.);

                    let character_spacing = Text {
                        extra_character_spacing: 16.,
                        ..Text::basic("Hello, World", &font, 32.)
                    };

                    let word_spacing = Text {
                        extra_word_spacing: 16.,
                        ..Text::basic("Hello, World", &font, 32.)
                    };

                    let both = Text {
                        extra_character_spacing: 16.,
                        extra_word_spacing: 16.,
                        ..Text::basic("Hello, World", &font, 32.)
                    };

                    content
                        .add(&normal.debug(0).show_max_width())?
                        .add(&character_spacing.debug(1).show_max_width())?
                        .add(&word_spacing.debug(2).show_max_width())?
                        .add(&both.debug(3).show_max_width())?;

                    None
                },
            });
        });
        assert_binary_snapshot!(".pdf", bytes);
    }

    #[test]
    fn test_truetype_soft_hyphen() {
        let mut params = TestElementParams::breakable();
        params.width.expand = false;

        let bytes = test_element_bytes(params, |mut callback| {
            let font = TruetypeFont::new(callback.pdf(), FONT);

            callback.call(&Column {
                gap: 12.,
                collapse: false,
                content: |content| {
                    let a = Text::basic("Hello\u{00AD}Wrld", &font, 32.);
                    let b = Text::basic("A Hello\u{00AD}Wrld", &font, 32.);
                    let c = Text::basic("A\u{00A0}Hello\u{00AD}Wrld", &font, 32.);
                    let d = Text::basic("Hello\u{00AD}Wrld\u{00AD}", &font, 32.);

                    content
                        .add(&Padding::right(100., a.debug(0).show_max_width()))?
                        .add(&Padding::right(120., b.debug(0).show_max_width()))?
                        .add(&Padding::right(120., c.debug(0).show_max_width()))?
                        .add(&Padding::right(20., d.debug(0).show_max_width()))?;

                    None
                },
            });
        });
        assert_binary_snapshot!(".pdf", bytes);
    }

    // #[test]
    // fn test_text() {
    //     // A fake document for adding the font to.
    //     let doc = PdfDocument::empty("i contain a font");

    //     let font = BuiltinFont::helvetica(&doc);

    //     let text_element = Text {
    //         ..Text::basic("i am a line\nso am i", &font, 12.)
    //     };

    //     let element = ElementProxy {
    //         before_draw: &|ctx: &mut DrawCtx| {
    //             // These seem to be stored in a map by name and when drawing a font with the same
    //             // needs to exist in the document being drawn on.
    //             ctx.pdf
    //                 .document
    //                 .add_builtin_font(printpdf::BuiltinFont::Helvetica)
    //                 .unwrap();
    //         },
    //         ..ElementProxy::new(text_element)
    //     };

    //     for mut output in (ElementTestParams {
    //         first_height: 4.,
    //         full_height: 5.,
    //         ..Default::default()
    //     })
    //     .run(&element)
    //     {
    //         if let Some(ref mut b) = output.breakable {
    //             b.assert_break_count(if output.first_height == 4. { 2 } else { 1 });
    //         }

    //         output.assert_size(ElementSize {
    //             width: Some(output.width.constrain(19.291312152)),

    //             // Note: I'm not sure this line height is correct. When running the same test with
    //             // Nimbus Sans L, which is supposed to be fully metrically compatible with
    //             // helvetica (at least according to the readme in
    //             // https://git.ghostscript.com/?p=user/tor/urw-base-12.git), the height ends up
    //             // being slightly more. On the x axis it matches exactly though. It's possible that
    //             // the bounding box in afm is not meant to be the equivalent of ascent + descent +
    //             // line gap in ttf. NimbusSans-Regular.afm from the following repo has the same
    //             // numbers https://git.ghostscript.com/?p=urw-core35-fonts.git as the Adobe one
    //             // and when running with NimbusSans-Regular.ttf from that repo the numbers are much
    //             // closer, which is reassuring (that one uses 2048 units per em so that should
    //             // explain the slight difference). The numbers for that one are 19.293924914062497
    //             // and 4.86792298828125.
    //             height: Some(4.893736415999999 * if output.breakable.is_some() { 1. } else { 2. }),
    //         });
    //     }
    // }
}
