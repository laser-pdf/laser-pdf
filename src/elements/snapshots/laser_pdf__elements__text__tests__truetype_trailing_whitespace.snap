---
source: src/elements/text.rs
expression: bytes
extension: pdf
snapshot_kind: binary
---
